### PR TITLE
Siamese Neural Networks on the Classification of Melanoma - 48191889

### PR DESCRIPTION
**Pull request by koningmalik412**
By Ainul Malik Zaidan Ismail - 48191889

Implements a module to train and test a Siamese Neural Network model to classify benign or malignant cases of melanoma based on the [Kaggle ISIC 2020 Challenge Dataset](https://challenge2020.isic-archive.com/).

Provides the following modules:
- `dataset.py`: Prepares the ISIC Dataset as a DataLoader for training and testing of the SNN.
- `modules.py`: Defines the architecture of the Siamese Neural Network, alongside methods for passing images through the twin channel architecture.
- `train.py`: Trains the SNN based on the training DataLoader. Returns the model alongside a loss plot of the training process.
- `predict.py`: Performs evaluation of the prediction performance of the SNN on unseen data of the training set and returns evaluation metrics.

### CODE REVIEW

**Comment by hanemma7moud** [comment]
### This is an initial inspection, no action is required at this point
Recognition Problem : total : **18**

Solves problem: The solution is well-suited to the problem (**3**).
+Triplet loss is utilized, which is suitable for the problem.
+Class imbalance is addressed through a mixed sampling approach that includes undersampling benign cases.
- only training loss plot is provided.
- no discriminative embeddings to visualize and evaluate the triplet loss results
- the reported performance is confusing ACC = 98.23%, Sensitivity = 98.87%, and specificity = 97.59%

Implementation functions as intended (**3**)

Good design: Well-designed (**1**)

Commenting: Clear and sufficient comments throughout the code. (**1**)

Difficulty: Hard (**10**)

Note:
**The confusion matrix shows a higher number of malignant samples** (1,855) than expected based on the reported dataset split,  where only around 467 malignant images should be in the 20% test set. This suggests that the evaluation might not reflect the original 80/20 split, or additional factors affected the test sample distribution, leading to inconsistencies in the reported results.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. Has data splits.</td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard.  Siamese network (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Some/Adequate Meaningful commit messages. </td><td>-1</td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td>-2</td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. Embeddings, loss curevs missing.</td><td>-2</td></tr>
	<tr><td></td><td>Model/technical explanation :Acceptable. </td><td>-2</td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-8</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Cant merge because of gitignore files. Please remove to allow merge, doesn't affect grade.
